File: ecommerce_integrations/shopify/report/shopify_inventory_mismatch/shopify_inventory_mismatch.js
// Copyright (c) 2026, Frappe and contributors
// For license information, please see LICENSE

frappe.query_reports["Shopify Inventory Mismatch"] = {
	filters: [
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			description: __(
				"Optional. Only mapped Shopify warehouses are compared. Leave blank for all mapped warehouses."
			),
			get_query: function () {
				return {
					query: "ecommerce_integrations.shopify.report.shopify_inventory_mismatch.shopify_inventory_mismatch.mapped_warehouse_query",
				};
			},
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "only_mismatches",
			label: __("Only Mismatches"),
			fieldtype: "Check",
			default: 1,
		},
	],

	get_datatable_options(options) {
		return Object.assign(options, {
			checkboxColumn: true,
		});
	},

	onload: function (report) {
		report.page.add_inner_button(__("Reconcile Selected"), () => {
			const selected = get_selected_rows(report);
			if (!selected.length) {
				frappe.msgprint(__("Select at least one row."));
				return;
			}
			frappe.call({
				method: "ecommerce_integrations.shopify.inventory.reconcile_inventory_rows",
				args: { rows: selected },
				freeze: true,
			});
		});

		report.page.add_inner_button(__("Reconcile All Mismatches On Page"), () => {
			const rows = get_mismatch_rows(report);
			if (!rows.length) {
				frappe.msgprint(__("No mismatch rows on this page."));
				return;
			}
			frappe.confirm(
				__("Push ERPNext qty to Shopify for {0} mismatch row(s)?", [rows.length]),
				() => {
					frappe.call({
						method: "ecommerce_integrations.shopify.inventory.reconcile_inventory_rows",
						args: { rows },
						freeze: true,
					});
				}
			);
		});

		report.page.add_inner_button(__("Push All Stock Now"), () => {
			frappe.confirm(
				__(
					"Push ERPNext stock for all mapped warehouses to Shopify now? This runs in the background."
				),
				() => {
					frappe.call({
						method: "ecommerce_integrations.shopify.inventory.push_inventory_to_shopify_now",
						freeze: true,
					});
				}
			);
		});
	},
};

function get_report_data(report) {
	return report.data || (frappe.query_report && frappe.query_report.data) || [];
}

function get_selected_rows(report) {
	const indices =
		report.datatable && report.datatable.rowmanager
			? report.datatable.rowmanager.getCheckedRows()
			: [];
	const data = get_report_data(report);
	return indices.map((i) => data[i]).filter(Boolean);
}

function get_mismatch_rows(report) {
	return get_report_data(report).filter((r) => {
		if (!r || r.status === "Not Found") {
			return false;
		}
		if (r.status === "Mismatch") {
			return true;
		}
		return cint(r.difference) !== 0;
	});
}

function cint(v) {
	const n = parseInt(v, 10);
	return Number.isNaN(n) ? 0 : n;
}

File: ecommerce_integrations/shopify/report/shopify_inventory_mismatch/shopify_inventory_mismatch.py
# Copyright (c) 2026, Frappe and contributors
# For license information, please see LICENSE

import frappe
from frappe import _
from frappe.desk.reportview import get_match_cond

from ecommerce_integrations.shopify.inventory import get_inventory_mismatch_data


def execute(filters=None):
	columns = get_columns()
	data = get_inventory_mismatch_data(filters)
	return columns, data


def get_columns():
	return [
		{
			"label": _("Item"),
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 160,
		},
		{
			"label": _("Ecommerce Item"),
			"fieldname": "ecom_item",
			"fieldtype": "Link",
			"options": "Ecommerce Item",
			"width": 140,
		},
		{
			"label": _("Variant Id"),
			"fieldname": "variant_id",
			"fieldtype": "Data",
			"width": 120,
		},
		{
			"label": _("Warehouse"),
			"fieldname": "warehouse",
			"fieldtype": "Link",
			"options": "Warehouse",
			"width": 150,
		},
		{
			"label": _("Shopify Location Id"),
			"fieldname": "shopify_location_id",
			"fieldtype": "Data",
			"width": 130,
		},
		{
			"label": _("ERP Available"),
			"fieldname": "erp_available",
			"fieldtype": "Int",
			"width": 110,
		},
		{
			"label": _("Reserved Qty"),
			"fieldname": "reserved_qty",
			"fieldtype": "Int",
			"width": 100,
		},
		{
			"label": _("Shopify Available"),
			"fieldname": "shopify_available",
			"fieldtype": "Int",
			"width": 130,
		},
		{
			"label": _("Difference (ERP − Shopify)"),
			"fieldname": "difference",
			"fieldtype": "Int",
			"width": 160,
		},
		{
			"label": _("Last Inventory Synced On"),
			"fieldname": "inventory_synced_on",
			"fieldtype": "Datetime",
			"width": 160,
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 100,
		},
	]


@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def mapped_warehouse_query(doctype, txt, searchfield, start, page_len, filters):
	"""Link query: only warehouses mapped on Shopify Setting."""
	setting = frappe.get_cached_doc("Shopify Setting")
	mapped = list(setting.get_erpnext_to_integration_wh_mapping().keys())
	if not mapped:
		return []

	txt = f"%{txt}%"
	return frappe.db.sql(
		f"""
		select name, warehouse_name
		from `tabWarehouse`
		where name in %(mapped)s
			and is_group = 0
			and ifnull(disabled, 0) = 0
			and (name like %(txt)s or warehouse_name like %(txt)s)
			{get_match_cond(doctype)}
		order by name
		limit %(start)s, %(page_len)s
		""",
		{"mapped": mapped, "txt": txt, "start": start, "page_len": page_len},
	)
